snuba/clickhouse/native.py

```diff
@@ -148,6 +150,7 @@ def __transform_result(self, result, with_totals: bool) -> Result:
         else:
             result = {"data": data, "meta": meta}
 
+        result["applied_sampling_rate"] = sampling_rate
```

Comment:
> I am not familiar on whether there will be any change in the structure of the response so I am not sure this is the right place to put the sampling rate in the response.

I don't think so. I will explain my reasoning below.

> I am open to move it out of the result object. The only requirement we have here is that the applied sampling rate must be in the response and it should not be possible to remove it (so it cannot be in the debug data).

This makes sense to me.

Since the `Reader` object does not choose the sampling rate (it is just executing the query as defined by the `Query` object and the method parameters), I don't think it belongs in the `Result` data structure that is returned by the `Reader`. I see the responsibilities of the `Reader` primarily as the gateway to stateful data (as defined by the query), including the column metadata from the ClickHouse system catalog. I think that the sampling rate exists outside of this scope. I also can't envision a scenario where it the sampling rate varies based on the `Reader` implementation.

Instead, I think this would be a better fit as part of `RawQueryResult` as a new attribute. (I am not sure that this is necessarily the _best_ place for it, but I do think it is a _better_ place for it.) From there it should be reasonably straightforward to add it to the response body.

(Note: I am making the assumption here that the sampling rate is constant once calculated which seems to me like a reasonable assumption, but I haven't read the rest of the proposed change yet.)

Reply:
```
 I am making the assumption here that the sampling rate is constant once calculated which seems to me like a reasonable assumption, but I haven't read the rest of the proposed change yet.
```
Yes this assumption is fair, in fact I added a specific method in ClickhouseQuery to make it harder to add code that changes the sampling rate while formatting (if something did, I think they deserve the consequences) while we do not have a Clickhouse AST.

About RawQueryResult.
Make sense to me, my reasoning was that Result was tied to the core part of the response while RawQueryResult had additional debug/metadata/stats, but since the sampling rate is not produced by the Reader it makes sense to me to move it.
As a followup, do you already have in mind how to format the stats data coming from clickhouse (bytes read, rows read, etc.) if we add them in the response? I wonder if the sampling rate should be a top level field in the response or part of a more complex metadata structure. I feel it does not really belong with any stats data since it is a parameter we provide and not something related with the performance of the query so maybe a top level field makes sense.

Reply:
> As a followup, do you already have in mind how to format the stats data coming from clickhouse (bytes read, rows read, etc.) if we add them in the response?

I'm not quite sure yet. I do think these should come from the `Result` structure somewhere. We should certainly include this in the query log, but I'm not sure that we should add them to the response.

> I wonder if the sampling rate should be a top level field in the response or part of a more complex metadata structure. I feel it does not really belong with any stats data since it is a parameter we provide and not something related with the performance of the query so maybe a top level field makes sense.

This depends on functionality that doesn't exist yet, but when there is a split between the Snuba query and the storage (ClickHouse) query, what if we returned the Snuba query (after processing, but before translating it to a storage query — essentially after all defaults were filled in) as part of the response? The user could just pull the actual value out of the `sample` clause of the response at that point. (I still haven't read through the rest of the PR in detail yet so making some assumptions how this works, will follow up if I realize this doesn't make sense.)

This way users would know exactly with what parameters the query was run with, and would be able to rerun the exact same query at a later date (assuming all processors at that stage are idempotent, and ignoring database state changes.) I think that this would basically be what `sql` in the response body was supposed to be before the introduction of the splitter.

Until that point, I think it makes sense to put this at the top level. I agree that I don't think it belongs with timing data or any of the other stats.